lib/options.go

```diff
@@ -44,6 +46,117 @@ const DefaultScenarioName = "default"
 // nolint: gochecknoglobals
 var DefaultSummaryTrendStats = []string{"avg", "min", "med", "max", "p(90)", "p(95)"}
 
+// DNSConfig is the DNS resolver configuration.
+type DNSConfig struct {
+	// If positive, defines how long DNS lookups should be returned from the cache.
+	TTL null.String `json:"ttl"`
+	// Strategy to use when picking a single IP if more than one is returned for a host name.
+	Strategy NullDNSStrategy `json:"strategy"`
+	// FIXME: Valid is unused and is only added to satisfy some logic in
+	// lib.Options.ForEachSpecified(), otherwise it would panic with
+	// `reflect: call of reflect.Value.Bool on zero Value`.
+	Valid bool
+}
+
+// DNSStrategy is the strategy to use when picking a single IP if more than one
+// is returned for a host name.
+//go:generate enumer -type=DNSStrategy -transform=kebab -trimprefix DNS -output dns_strategy_gen.go
+type DNSStrategy uint8
+
+// NullDNSStrategy is a nullable wrapper around DNSStrategy, required for the
+// current configuration system.
+type NullDNSStrategy struct {
+	DNSStrategy
+	Valid bool
+}
+
+const (
+	// DNSFirst returns the first IP from the response.
+	DNSFirst DNSStrategy = iota + 1
+	// DNSRoundRobin rotates the IP returned on each lookup.
+	DNSRoundRobin
+	// DNSRandom returns a random IP from the response.
+	DNSRandom
+)
+
+// DefaultDNSConfig returns the default DNS configuration.
+func DefaultDNSConfig() DNSConfig {
+	return DNSConfig{
+		TTL:      null.NewString("5m", false),
+		Strategy: NullDNSStrategy{DNSRandom, false},
+	}
+}
+
+// String implements fmt.Stringer.
+func (c DNSConfig) String() string {
+	out := make([]string, 0, 2)
+	out = append(out, fmt.Sprintf("ttl=%s", c.TTL.String))
+	out = append(out, fmt.Sprintf("strategy=%s", c.Strategy.String()))
+	return strings.Join(out, ",")
+}
+
+// MarshalJSON implements json.Marshaler.
+func (c DNSConfig) MarshalJSON() ([]byte, error) {
+	var s string
+	if c.Strategy.IsADNSStrategy() {
+		s = c.Strategy.String()
+	}
+	return json.Marshal(struct {
+		TTL      string `json:"ttl"`
+		Strategy string `json:"strategy"`
+	}{
+		TTL:      c.TTL.String,
+		Strategy: s,
+	})
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (c *DNSConfig) UnmarshalJSON(data []byte) error {
+	params := make(map[string]interface{})
+	if err := json.Unmarshal(data, &params); err != nil {
+		return err
+	}
+	return c.unmarshal(params)
+}
+
+// UnmarshalText implements encoding.TextUnmarshaler.
+func (c *DNSConfig) UnmarshalText(text []byte) error {
+	if string(text) == DefaultDNSConfig().String() {
+		*c = DefaultDNSConfig()
+		return nil
+	}
+	params, err := strvals.Parse(string(text))
```

Comment:
Maybe... But I don't want to introduce any potential parsing issues if `strvals` already handles it. Plus `strvals` is already used in the Kafka and InfluxDB configs, so it's not like we're adding a new dependency with this.

So I'd rather do this once we remove `strvals` with that bespoke version you guys were considering.

Reply:
> any potential parsing issues if strvals already handles it. 

you are hitting the potential parsing issues ... because you have an `interface{}` instead of a string ... but w/e
> Plus strvals is already used in the Kafka and InfluxDB configs, so it's not like we're adding a new dependency with this.

yes ... and using it in one more place makes it that much harder to remove it ;) 

again this is nitpicking, I am not saying this is a required thing ... just pointing out that the usage of `strvals` in this concrete example makes other code necessary that wouldn't have been and arguably adds zero value IMO